staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -469,12 +477,38 @@ func (s unstructuredNegotiatedSerializer) SupportedMediaTypes() []runtime.Serial
 	}
 }
 
+type CustomResourceEncoderDecoder struct {
```

Comment:
I think @liggitt commented the same somewhere: this should go into a converter instead of being done during decoding/encoding.

Reply:
yep, that's what I am going to figure out tomorrow. I just tried to make it work today and looked into doing it in converters stack but didn't find a good way to do it. Any suggestion is welcome.

Reply:
We have the UnstructuredObjectConverter. Any mutation of the GroupVersion fits better there than in the serialization.

What is a bit odd today is that for unstructure we keep the apiVersion set all the time, while with native resource we do not keep it when converting to the internal version. Not sure we can and want to rebuild this behaviour here with unstructured. @liggitt opinions?

Reply:
> I think @liggitt commented the same somewhere: this should go into a converter instead of being done during decoding/encoding.

Thirded.  It's a few lines up in the file.

Reply:
The problem is that converter is not called for CRDs. I tried it first.

Reply:
> The problem is that converter is not called for CRDs. I tried it first.

Then that's the point to update in this PR to position us correctly and preserve the conversion abstractions

Reply:
I looked into this a little more and look like the Converter (and the UnsafeConverter) is only use for label conversion and it is the serializer that used for conversion. Do you guys think that is not the right conclusion?
Our internal Objects have a codec factory that creates the serializer that also do conversion based on Schema.Convert method. We can construct our CRD stack like that (instead of creating our serializer) but the problem is Schema.Convert will not do conversions on unstructured objects and also not all unstructured objects are CRDs.
Me and @sttts discussed the idea of wrapping unstructured object into a CustomResource object and do conversion on that. I went deeper into this solution and find it a decent solution as it will let us use current conversion stack and enable us to use external links in CRD schema later (e.g. referencing a Pod and being able to convert between Pod versions, not a goal now but good to have our options open). I am going to implement it and send it for another round of review.

Reply:
> We can construct our CRD stack like that (instead of creating our serializer) but the problem is Schema.Convert will not do conversions on unstructured objects and also not all unstructured objects are CRDs

The serializer and convertor for CRD will always be a custom implementation (don't try to make the existing scheme convertor implementations CRD or unstructured-aware). CRD storage should call its convertor on the way in and out of etcd in a similar way to how normal storage calls its convertor.

Reply:
> wrapping unstructured object into a CustomResource object and do conversion on that

This seems like a bad idea to me… that's how TPRs tried to treat custom resource data and it had bad side effects. The go struct based conversions use reverse lookups based on the go types to determine versions and convertibility… staying with unstructured and implementing conversions based on the apiVersion contained in the object seems much less likely to cause problems elsewhere

Reply:
> The serializer and convertor for CRD will always be a custom implementation

The scheme entry methods are Unstructured-aware and refuse to do anything sensible. I will dive into the code this morning to see how we can inject our converter properly.

Reply:
Here I add a NoConversion converter for CRs: https://github.com/mbohlool/kubernetes/pull/2

The non-trivial change is https://github.com/mbohlool/kubernetes/pull/2/files#diff-2dc3c4b11095961559646d293d1f6d16R170 in the default versioned serializer: it was special casing Unstructured in ConvertToVersion, skipping all conversion logic. Our default ObjectTyper has support for Unstructured. So the default code path should work.